Add docs about unit tests practices #1168 (#1272)
CGI-FR/IoT-Hub-Portal: docs/dev-guide/testing/unit-tests-common-practices.md

# Unit Tests Common Practices

## Naming Conventions

### Test class

Test class should follow the naming convention **`[ClassUnderTest]Tests`**.

Example: The test class for a class named ProductController should be `ProductControllerTests`:

```csharp
[TestFixture]
public class ProductControllerTests
{
    ...
}
```

### Test method

Test method should follow the naming convention **`[MethodUnderTest]_[BehaviourToTest]_[ExpectedResult]`**.

Example: For a method called GetProduct, the behaviour that we want to test is that it should return an exsiting project.
The name of the test should be `GetProduct_ProductExist_ProductReturned`:

```csharp
[Test]
public async Task GetProduct_ProductExist_ProductReturned()
{
    ...
}
```

## Unit Test Skeleton: Three Steps/Parts

> A unit test should be designed/cut into three steps:

Reviewer: ```suggestion
A unit test should be divided into three steps:
```

> 1. Arrange: The first part where input/expected data are defined
> 2. Act: The second part where the behavior under test is executed
> 3. Assert: The third and last part where assertions are done

Reviewer: ```suggestion
1. Arrange: The first part where the input/expected data is defined
2. Act: The second part where the behavior under test is executed
3. Assert: The third and final part where assertions are made
```

> These three parts are visually defined with comments so that unit tests can be humain readeable:

Reviewer: ```suggestion
These three parts are visually defined with comments so that unit tests are humanly comprehensible:
```

```csharp
[Test]
public async Task GetProduct_ProductExist_ProductReturned()
{
    // Arrange
    var productId = Guid.NewGuid().ToString();
    var expectedProduct = new Product
    {
        Id = productId
    };

    // Act
    var product = this.productService.GetProduct(productId);

    // Asset
    _ = product.Should().BeEquivalentTo(expectedProduct);
}
```

!!! Tip

>     On IoT Hub Portal, we use the library [fluentassertions](https://github.com/fluentassertions/fluentassertions) on unit tests for natural/humain readeable assertions.

Reviewer: ```suggestion
    On the IoT Hub portal, we use the [fluentassertions] library (https://github.com/fluentassertions/fluentassertions) for unit tests for natural/human reusable assertions.
```

## Mock

> A unit test should only test its dedicated layer. Any lower layer that requires/interact with external resources should be mocked to make sure that unit tests are idempotents.

Reviewer: ```suggestion
A unit test should only test its assigned layer. Any lower layer that requires/interacts with external resources should be mocked to ensure that the unit tests are idempotent.
```

!!! note

>     For example, we want to implement unit tests on a controller that requires three services. Each service depends on others services/repositories/http clients that require external resources like databases, APIs...
>     Any execution of unit tests that require on these external resources can be altered (not idempotent) because they depend on the uptime, the data of these resources.

Reviewer: ```suggestion
    Example: We want to implement unit tests for a controller that requires three services. Each service depends on other services/repositories/http clients that need external resources like databases, APIs.... 
    Any execution of unit tests that depend on these external resources can be modified (not idempotent) because they depend on the uptime and data of these resources.
```

> On IoT Hub Portal, we use the library [Moq](https://github.com/moq/moq4) for mocking within unit tests:

Reviewer: ```suggestion
On the IoT Hub portal, we use the library [Moq](https://github.com/moq/moq4) for mocking within unit tests:
```
